airflow/providers/amazon/aws/hooks/elasticache.py

```diff
@@ -0,0 +1,259 @@
+
+        while num_retries < max_retries:
+            stop_poking, status = self.should_stop_poking(replication_group_id=replication_group_id)
+
+            self.log.info('Status : %s', status)
+
+            if stop_poking:
+                self.log.info('Received signal to stop poking. Current status : "%s"', status)
+                break
+
+            num_retries += 1
+
+            self.log.info('Poke retry: %s. Sleep time: %s. Sleeping...', num_retries, sleep_time)
+
+            sleep(sleep_time)
+
+            sleep_time = sleep_time * exponential_back_off_factor
+
+        if status != self.STATUS_AVAILABLE:
+            self.log.warning('Replication group is not available. Current status : "%s"', status)
+
+            return False
+
+        return True
+
+    def wait_for_deletion(
+        self,
+        replication_group_id,
+        initial_sleep_time=None,
+        exponential_back_off_factor=None,
+        max_retries=None
+    ):
+        """
+        Helper for deleting a replication group ensuring it is either deleted or can't be deleted
+        :param replication_group_id: ID of replication to delete
+        :param max_retries: Max tries for checking availability of replication group
+        :param exponential_back_off_factor: Factor for deciding next sleep time
+        :param initial_sleep_time: Initial sleep time in second
+        :return: Response from ElastiCache delete replication group API and flag to identify if deleted or not
+        """
+        response = None
+        deleted = False
+        sleep_time = initial_sleep_time or self.initial_poke_interval
+        exponential_back_off_factor = exponential_back_off_factor or self.exponential_back_off_factor
+        max_retries = max_retries or self.max_retries
+        num_retries = 0
+
+        while not deleted and num_retries < max_retries:
+            try:
+                response = self.delete_replication_group(replication_group_id=replication_group_id)
+
+                self.log.info('Replication group with ID : %s : is being deleted...', replication_group_id)
+
+            except self.conn.exceptions.ReplicationGroupNotFoundFault:
+                self.log.info("Replication group with ID : '%s' does not exist", replication_group_id)
+
+                deleted = True
+
+            except self.conn.exceptions.InvalidReplicationGroupStateFault as exp:
+                message = exp.response['Error']['Message']
+
+                self.log.info('Received message : %s', message)
+
+                if re.search(self.RESPONSE_NOT_VALID_STATE, message) \
+                        or re.search(self.RESPONSE_MODIFYING_STATE, message):
+
+                    self.log.info(
+                        'Replication group with ID : %s : is not in valid state to be deleted',
+                        replication_group_id
+                    )
+
+                elif re.search(self.RESPONSE_DELETING_STATE, message):
+                    self.log.info(
```

Comment:
Wait, we try and delete it every time around the loop?

That seems.... odd. Why not 1) issue delete request, 2) describe/sleep until it is eventually deleted.

Reply:
Because we delete from any state but only `available`, hence I issued `delete` each time. I have changed it to get status first and then only issue a `delete` when status is `available`.